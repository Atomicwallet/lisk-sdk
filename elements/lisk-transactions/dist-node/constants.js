"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_TRANSFER_ASSET_DATA_LENGTH = exports.MAX_PUBLIC_KEY_LENGTH = exports.MAX_MULTISIG_SIGNATURES = exports.UNCONFIRMED_MULTISIG_TRANSACTION_TIMEOUT = exports.UNCONFIRMED_TRANSACTION_TIMEOUT = exports.MAX_TRANSACTION_AMOUNT = exports.MAX_TRANSACTION_ID = exports.MAX_ADDRESS_NUMBER = exports.EPOCH_TIME_SECONDS = exports.EPOCH_TIME_MILLISECONDS = exports.EPOCH_TIME = exports.BYTESIZES = exports.USERNAME_MAX_LENGTH = exports.DAPP_FEE = exports.MULTISIGNATURE_MIN_KEYSGROUP = exports.MULTISIGNATURE_MAX_KEYSGROUP = exports.MULTISIGNATURE_MIN_LIFETIME = exports.MULTISIGNATURE_MAX_LIFETIME = exports.MULTISIGNATURE_FEE = exports.VOTE_FEE = exports.DELEGATE_FEE = exports.SIGNATURE_FEE = exports.OUT_TRANSFER_FEE = exports.IN_TRANSFER_FEE = exports.TRANSFER_FEE = exports.FIXED_POINT = void 0;
exports.FIXED_POINT = 10 ** 8;
exports.TRANSFER_FEE = exports.FIXED_POINT * 0.1;
exports.IN_TRANSFER_FEE = exports.FIXED_POINT * 0.1;
exports.OUT_TRANSFER_FEE = exports.FIXED_POINT * 0.1;
exports.SIGNATURE_FEE = exports.FIXED_POINT * 5;
exports.DELEGATE_FEE = exports.FIXED_POINT * 25;
exports.VOTE_FEE = exports.FIXED_POINT * 1;
exports.MULTISIGNATURE_FEE = exports.FIXED_POINT * 5;
exports.MULTISIGNATURE_MAX_LIFETIME = 72;
exports.MULTISIGNATURE_MIN_LIFETIME = 1;
exports.MULTISIGNATURE_MAX_KEYSGROUP = 15;
exports.MULTISIGNATURE_MIN_KEYSGROUP = 1;
exports.DAPP_FEE = exports.FIXED_POINT * 25;
exports.USERNAME_MAX_LENGTH = 20;
exports.BYTESIZES = {
    TYPE: 1,
    TIMESTAMP: 4,
    MULTISIGNATURE_PUBLICKEY: 32,
    RECIPIENT_ID: 8,
    AMOUNT: 8,
    SIGNATURE_TRANSACTION: 64,
    SECOND_SIGNATURE_TRANSACTION: 64,
    DATA: 64,
};
exports.EPOCH_TIME = new Date(Date.UTC(2016, 4, 24, 17, 0, 0, 0));
exports.EPOCH_TIME_MILLISECONDS = exports.EPOCH_TIME.getTime();
const MS_FACTOR = 1000;
exports.EPOCH_TIME_SECONDS = Math.floor(exports.EPOCH_TIME.getTime() / MS_FACTOR);
const MAX_EIGHT_BYTE_NUMBER = '18446744073709551615';
exports.MAX_ADDRESS_NUMBER = MAX_EIGHT_BYTE_NUMBER;
exports.MAX_TRANSACTION_ID = MAX_EIGHT_BYTE_NUMBER;
exports.MAX_TRANSACTION_AMOUNT = '9223372036854775807';
exports.UNCONFIRMED_TRANSACTION_TIMEOUT = 10800;
exports.UNCONFIRMED_MULTISIG_TRANSACTION_TIMEOUT = 10800 * 8;
exports.MAX_MULTISIG_SIGNATURES = 15;
exports.MAX_PUBLIC_KEY_LENGTH = 32;
exports.MAX_TRANSFER_ASSET_DATA_LENGTH = 64;
//# sourceMappingURL=constants.js.map