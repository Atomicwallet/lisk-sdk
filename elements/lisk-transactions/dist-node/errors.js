"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var TransactionError = (function (_super) {
    __extends(TransactionError, _super);
    function TransactionError(message, id, dataPath, actual, expected) {
        if (message === void 0) { message = ''; }
        if (id === void 0) { id = ''; }
        if (dataPath === void 0) { dataPath = ''; }
        var _this = _super.call(this) || this;
        _this.message = message;
        _this.name = 'TransactionError';
        _this.id = id;
        _this.dataPath = dataPath;
        _this.actual = actual;
        _this.expected = expected;
        return _this;
    }
    TransactionError.prototype.toString = function () {
        var defaultMessage = "Transaction: " + this.id + " failed at " + this.dataPath + ": " + this.message;
        var withActual = this.actual
            ? defaultMessage + ", actual: " + this.actual
            : defaultMessage;
        var withExpected = this.expected
            ? withActual + ", expected: " + this.expected
            : withActual;
        return withExpected;
    };
    return TransactionError;
}(Error));
exports.TransactionError = TransactionError;
var TransactionPendingError = (function (_super) {
    __extends(TransactionPendingError, _super);
    function TransactionPendingError(message, id, dataPath) {
        if (message === void 0) { message = ''; }
        if (id === void 0) { id = ''; }
        if (dataPath === void 0) { dataPath = ''; }
        var _this = _super.call(this, message) || this;
        _this.name = 'TransactionPendingError';
        _this.id = id;
        _this.dataPath = dataPath;
        return _this;
    }
    TransactionPendingError.prototype.toString = function () {
        return "Transaction: " + this.id + " failed at " + this.dataPath + ": " + this.message + " ";
    };
    return TransactionPendingError;
}(TransactionError));
exports.TransactionPendingError = TransactionPendingError;
exports.convertToTransactionError = function (id, errors) {
    if (!errors) {
        return [];
    }
    return errors.map(function (error) {
        return new TransactionError("'" + error.dataPath + "' " + error.message, id, error.dataPath);
    });
};
exports.convertToAssetError = function (id, errors) {
    if (!errors) {
        return [];
    }
    return errors.map(function (error) {
        return new TransactionError("'" + (error.dataPath || '.asset') + "' " + error.message, id, error.dataPath || '.asset');
    });
};
//# sourceMappingURL=errors.js.map