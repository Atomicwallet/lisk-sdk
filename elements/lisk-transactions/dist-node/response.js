"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponse = exports.Status = void 0;
var Status;
(function (Status) {
    Status[Status["FAIL"] = 0] = "FAIL";
    Status[Status["OK"] = 1] = "OK";
    Status[Status["PENDING"] = 2] = "PENDING";
})(Status = exports.Status || (exports.Status = {}));
const createResponse = (id, errors) => ({
    id,
    status: errors && errors.length > 0 ? Status.FAIL : Status.OK,
    errors: errors ? errors : [],
});
exports.createResponse = createResponse;
//# sourceMappingURL=response.js.map