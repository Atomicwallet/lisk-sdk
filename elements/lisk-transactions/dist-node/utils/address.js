"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lisk_cryptography_1 = require("@liskhq/lisk-cryptography");
var errors_1 = require("../errors");
exports.validateSenderIdAndPublicKey = function (id, senderId, senderPublicKey) {
    var actualAddress = lisk_cryptography_1.getAddressFromPublicKey(senderPublicKey);
    return senderId.toUpperCase() !== actualAddress.toUpperCase()
        ? new errors_1.TransactionError('`senderId` does not match `senderPublicKey`', id, '.senderId', actualAddress, senderId)
        : undefined;
};
//# sourceMappingURL=address.js.map