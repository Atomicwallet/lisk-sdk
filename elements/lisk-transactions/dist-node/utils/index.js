"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./address"), exports);
__exportStar(require("./create_base_transaction"), exports);
__exportStar(require("./get_transaction_bytes"), exports);
__exportStar(require("./get_transaction_hash"), exports);
__exportStar(require("./transaction_id"), exports);
__exportStar(require("./format"), exports);
__exportStar(require("./sign_and_validate"), exports);
__exportStar(require("./time"), exports);
__exportStar(require("./validation"), exports);
__exportStar(require("./verify"), exports);
__exportStar(require("./prepare_transaction"), exports);
__exportStar(require("./sign_raw_transaction"), exports);
//# sourceMappingURL=index.js.map