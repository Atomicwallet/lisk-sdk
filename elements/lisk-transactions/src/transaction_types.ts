/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { TransactionError } from './errors';
// tslint:disable readonly-keyword
export interface Account {
	readonly address: string;
	balance: bigint;
	missedBlocks: number;
	producedBlocks: number;
	publicKey: string | undefined;
	username: string | null;
	isDelegate: number;
	fees: bigint;
	rewards: bigint;
	voteWeight: bigint;
	nameExist: boolean;
	multiMin: number;
	multiLifetime: number;
	asset: object;
	votedDelegatesPublicKeys: string[];
	membersPublicKeys: string[];
	// tslint:disable-next-line:no-mixed-interface
	readonly toJSON: () => object;
}
// tslint:enable readonly-keyword
export interface Delegate {
	readonly username: string;
}

export interface TransactionJSON {
	readonly asset: object;
	readonly id?: string;
	readonly blockId?: string;
	readonly height?: number;
	readonly confirmations?: number;
	readonly senderPublicKey: string;
	readonly signature?: string;
	readonly signatures?: ReadonlyArray<string>;
	readonly signSignature?: string;
	readonly type: number;
	readonly receivedAt?: string;
	readonly networkIdentifier?: string;
	readonly nonce: string;
	readonly fee: string;
}

export interface IsValidResponse {
	readonly valid: boolean;
	readonly errors?: ReadonlyArray<TransactionError>;
}

export interface IsValidResponseWithError {
	readonly valid: boolean;
	readonly error?: TransactionError;
}

export interface IsVerifiedResponse {
	readonly verified: boolean;
	readonly errors?: ReadonlyArray<TransactionError>;
}
