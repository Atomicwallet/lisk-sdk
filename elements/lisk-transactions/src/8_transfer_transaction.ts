/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BaseTransaction, StateStore } from './base_transaction';
import { MAX_TRANSACTION_AMOUNT } from './constants';
import { TransactionError } from './errors';
import { verifyMinRemainingBalance } from './utils';
import { BaseTransactionInput } from './types';

export interface TransferAsset {
	readonly amount: bigint;
	readonly recipientAddress: Buffer;
	readonly data: string;
}

export const transferAssetSchema = {
	$id: 'lisk/transfer-transaction',
	title: 'Transfer transaction asset',
	type: 'object',
	required: ['amount', 'recipientAddress', 'data'],
	properties: {
		amount: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		recipientAddress: {
			dataType: 'bytes',
			fieldNumber: 2,
			minLength: 20,
			maxLength: 20,
		},
		data: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 0,
			maxLength: 64,
		},
	},
};

export class TransferTransaction extends BaseTransaction {
	public static TYPE = 8;
	public static ASSET_SCHEMA = transferAssetSchema;
	public readonly asset: TransferAsset;

	public constructor(transaction: BaseTransactionInput<TransferAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get(this.senderId);

		sender.balance -= this.asset.amount;
		store.account.set(sender.address, sender);
		const recipient = await store.account.getOrDefault(
			this.asset.recipientAddress,
		);

		recipient.balance += this.asset.amount;

		if (recipient.balance > BigInt(MAX_TRANSACTION_AMOUNT)) {
			errors.push(
				new TransactionError(
					'Invalid amount',
					this.id,
					'.amount',
					this.asset.amount.toString(),
				),
			);
		}

		// Validate minimum remaining balance
		const minRemainingBalanceError = verifyMinRemainingBalance(
			this.id,
			recipient,
			(this.constructor as typeof BaseTransaction).MIN_REMAINING_BALANCE,
		);
		if (minRemainingBalanceError) {
			errors.push(minRemainingBalanceError);
		}

		store.account.set(recipient.address, recipient);

		return errors;
	}
}
