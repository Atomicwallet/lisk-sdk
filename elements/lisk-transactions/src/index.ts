/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { ProofOfMisbehaviorTransaction } from './15_proof_of_misbehavior_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import { BaseTransaction, StateStore } from './base_transaction';
import * as constants from './constants';
import { convertToAssetError, convertToTransactionError, TransactionError } from './errors';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import {
	getSigningBytes,
	validateTransactionSchema,
	signTransaction,
	signMultiSignatureTransaction,
} from './transaction_helper';
import { Account, TransactionJSON } from './types';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
};

export {
	Account,
	BaseTransaction,
	getSigningBytes,
	StateStore,
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	UnlockTransaction,
	createResponse,
	ProofOfMisbehaviorTransaction,
	signTransaction,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	validateTransactionSchema,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
