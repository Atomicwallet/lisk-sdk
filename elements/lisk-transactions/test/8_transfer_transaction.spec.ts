/*
 * Copyright © 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { MAX_TRANSACTION_AMOUNT } from '../src/constants';
import { TransferTransaction } from '../src/8_transfer_transaction';
import { Account } from '../src/transaction_types';
import { Status } from '../src/response';
import { TransactionError } from '../src/errors';
import { defaultAccount, StateStoreMock } from './utils/state_store_mock';
import * as fixture from '../fixtures/transaction_network_id_and_change_order/transfer_transaction_validate.json';
import { BaseTransaction } from '../src';

describe('Transfer transaction class', () => {
	const validTransferTransaction = fixture.testCases[0].output;
	const validTransferAccount = fixture.testCases[0].input.account;
	let validTransferTestTransaction: TransferTransaction;
	let sender: Account;
	let recipient: Account;
	let store: StateStoreMock;

	beforeEach(async () => {
		validTransferTestTransaction = new TransferTransaction(
			validTransferTransaction,
		);
		sender = {
			...defaultAccount,
			...validTransferAccount,
			balance: BigInt('10000000000'),
			address: validTransferTestTransaction.senderId,
		};
		recipient = {
			...defaultAccount,
			...validTransferAccount,
			balance: BigInt('10000000000'),
			address: validTransferTestTransaction.asset.recipientId,
		};

		store = new StateStoreMock([sender, recipient]);

		jest.spyOn(store.account, 'cache');
		jest.spyOn(store.account, 'get');
		jest.spyOn(store.account, 'getOrDefault');
		jest.spyOn(store.account, 'set');
	});

	describe('#constructor', () => {
		it('should create instance of TransferTransaction', async () => {
			expect(validTransferTestTransaction).toBeInstanceOf(TransferTransaction);
		});

		it('should set transfer asset data', async () => {
			expect(validTransferTestTransaction.asset.data).toEqual(
				validTransferTestTransaction.asset.data,
			);
		});

		it('should set transfer asset amount', async () => {
			expect(validTransferTestTransaction.asset.amount.toString()).toEqual(
				validTransferTransaction.asset.amount,
			);
		});

		it('should set transfer asset recipientId', async () => {
			expect(validTransferTestTransaction.asset.recipientId).toEqual(
				validTransferTransaction.asset.recipientId,
			);
		});
	});

	describe('#verifyAgainstOtherTransactions', () => {
		it('should return a successful transaction response', async () => {
			const {
				id,
				status,
				errors,
			} = validTransferTestTransaction.verifyAgainstOtherTransactions([]);
			expect(id).toEqual(validTransferTransaction.id);
			expect(Object.keys(errors)).toHaveLength(0);
			expect(status).toEqual(Status.OK);
		});
	});

	describe('#assetToJSON', () => {
		it('should return an object of type transfer asset', async () => {
			expect(
				(validTransferTestTransaction.assetToJSON() as any).data,
			).toBeString();
		});
	});

	describe('#prepare', () => {
		it('should call state store', async () => {
			await validTransferTestTransaction.prepare(store);
			expect(store.account.cache).toHaveBeenCalledWith([
				{ address: validTransferTestTransaction.senderId },
				{ address: validTransferTestTransaction.asset.recipientId },
			]);
		});
	});

	describe('#validateAsset', () => {
		it('should return no errors with a valid transfer transaction', async () => {
			const errors = (validTransferTestTransaction as any).validateAsset();
			expect(Object.keys(errors)).toHaveLength(0);
		});

		it('should return error with invalid recipientId', async () => {
			const transferTransactionWithInvalidRecipientId = new TransferTransaction(
				{
					...validTransferTransaction,
					asset: {
						...validTransferTransaction.asset,
						recipientId: '123456',
					},
				},
			);
			const errors = (transferTransactionWithInvalidRecipientId as any).validateAsset();

			expect(errors[0]).toBeInstanceOf(TransactionError);
			expect(errors[0].message).toContain(
				'\'.recipientId\' should match format "address"',
			);
		});

		it('should return error if recipientId exceed uint64 limit', async () => {
			const transferTransactionWithInvalidRecipientId = new TransferTransaction(
				{
					...validTransferTransaction,
					asset: {
						...validTransferTransaction.asset,
						recipientId: '19961131544040416558L',
					},
				},
			);
			const errors = (transferTransactionWithInvalidRecipientId as any).validateAsset();

			expect(errors).toHaveLength(1);
			expect(errors[0]).toBeInstanceOf(TransactionError);
		});

		it('should return error if recipientId contains leading zeros', async () => {
			const transferTransactionWithInvalidRecipientId = new TransferTransaction(
				{
					...validTransferTransaction,
					asset: {
						...validTransferTransaction.asset,
						recipientId: '000123L',
					},
				},
			);
			const errors = (transferTransactionWithInvalidRecipientId as any).validateAsset();

			expect(errors).toHaveLength(1);
			expect(errors[0]).toBeInstanceOf(TransactionError);
		});

		it('should return error with invalid amount', async () => {
			const transferTransactionWithInvalidAmount = new TransferTransaction({
				...validTransferTransaction,
				asset: {
					...validTransferTransaction.asset,
					amount: '92233720368547758087823474829847337',
				},
			});
			const errors = (transferTransactionWithInvalidAmount as any).validateAsset();

			expect(errors[0]).toBeInstanceOf(TransactionError);
			expect(errors[0].message).toEqual(
				'Amount must be a valid number in string format.',
			);
			expect(errors[0].dataPath).toEqual('.asset.amount');
		});

		it('should return error with invalid asset', async () => {
			const transferTransactionWithInvalidAsset = new TransferTransaction({
				...validTransferTransaction,
				asset: {
					...validTransferTransaction.asset,
					data:
						'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
				},
			});
			const errors = (transferTransactionWithInvalidAsset as any).validateAsset();

			expect(errors[0]).toBeInstanceOf(TransactionError);
		});

		it('should return error if asset data containing null string', async () => {
			const transferTransactionWithValiddAsset = new TransferTransaction({
				...validTransferTransaction,
				asset: {
					...validTransferTransaction.asset,
					data: '\u0000hey:)',
				},
			});
			const errors = (transferTransactionWithValiddAsset as any).validateAsset();

			expect(errors).toHaveLength(1);
			expect(errors[0]).toBeInstanceOf(TransactionError);
		});

		it('should return error with asset data containing overflowed string', async () => {
			const transferTransactionWithInvalidAsset = new TransferTransaction({
				...validTransferTransaction,
				asset: {
					data:
						'o2ljg313lzzopdcilxcuy840qzdnmj21hfehd8u63k9jkifpsgxptegi56t8xos现',
				},
			});
			const errors = (transferTransactionWithInvalidAsset as any).validateAsset();

			expect(errors[0]).toBeInstanceOf(TransactionError);
		});
	});

	describe('#applyAsset', () => {
		it('should return no errors', async () => {
			const errors = (validTransferTestTransaction as any).applyAsset(store);

			expect(Object.keys(errors)).toHaveLength(0);
		});

		it('should call state store', async () => {
			await (validTransferTestTransaction as any).applyAsset(store);
			expect(store.account.get).toHaveBeenCalledWith(
				validTransferTestTransaction.senderId,
			);
			expect(store.account.set).toHaveBeenCalledWith(
				sender.address,
				expect.objectContaining({
					address: sender.address,
					publicKey: sender.publicKey,
				}),
			);
			expect(store.account.getOrDefault).toHaveBeenCalledWith(
				validTransferTestTransaction.asset.recipientId,
			);
			expect(store.account.set).toHaveBeenCalledWith(
				recipient.address,
				expect.objectContaining({
					address: recipient.address,
					publicKey: recipient.publicKey,
				}),
			);
		});

		it('should return error when sender balance is insufficient', async () => {
			store.account.set(sender.address, {
				...sender,
				balance: BigInt(10000000),
			});
			const errors = await (validTransferTestTransaction as any).applyAsset(
				store,
			);
			expect(errors).toHaveLength(1);
			expect(errors[0].message).toContain(
				'Account does not have enough minimum remaining LSK',
			);
		});

		it('should return error when recipient balance is over maximum amount', async () => {
			store.account.set(recipient.address, {
				...recipient,
				balance: BigInt(MAX_TRANSACTION_AMOUNT),
			});
			const errors = await (validTransferTestTransaction as any).applyAsset(
				store,
			);
			expect(errors[0].message).toEqual('Invalid amount');
		});

		it('should return error when recipient balance is below minimum remaining balance', async () => {
			store.account.set(recipient.address, {
				...recipient,
				balance:
					-validTransferTestTransaction.asset.amount +
					BaseTransaction.MIN_REMAINING_BALANCE -
					BigInt(1),
			});
			const errors = await (validTransferTestTransaction as any).applyAsset(
				store,
			);
			expect(errors[0].message).toContain(
				'Account does not have enough minimum remaining LSK',
			);
		});
	});

	describe('#undoAsset', () => {
		it('should call state store', async () => {
			await (validTransferTestTransaction as any).undoAsset(store);
			expect(store.account.get).toHaveBeenCalledWith(
				validTransferTestTransaction.senderId,
			);

			expect(store.account.set).toHaveBeenCalledWith(
				sender.address,
				expect.objectContaining({
					address: sender.address,
					publicKey: sender.publicKey,
				}),
			);
			expect(store.account.getOrDefault).toHaveBeenCalledWith(
				validTransferTestTransaction.asset.recipientId,
			);
			expect(store.account.set).toHaveBeenCalledWith(
				recipient.address,
				expect.objectContaining({
					address: recipient.address,
					publicKey: recipient.publicKey,
				}),
			);
		});

		it('should return error when sender balance is over maximum amount', async () => {
			store.account.set(sender.address, {
				...sender,
				balance: BigInt(MAX_TRANSACTION_AMOUNT),
			});
			const errors = await (validTransferTestTransaction as any).undoAsset(
				store,
			);
			expect(errors[0].message).toEqual('Invalid amount');
		});
	});
});
