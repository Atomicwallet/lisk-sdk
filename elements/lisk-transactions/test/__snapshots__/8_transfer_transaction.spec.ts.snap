// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transfer transaction class #sign should have two signatures for a multisignature account used as 2nd passphrase account 1`] = `
Array [
  Object {
    "data": Array [
      128,
      74,
      23,
      176,
      114,
      103,
      246,
      13,
      35,
      76,
      206,
      5,
      9,
      5,
      197,
      119,
      172,
      112,
      95,
      46,
      139,
      51,
      19,
      41,
      104,
      208,
      187,
      26,
      160,
      54,
      12,
      182,
      15,
      73,
      11,
      171,
      12,
      12,
      200,
      86,
      216,
      240,
      62,
      135,
      170,
      183,
      65,
      179,
      113,
      67,
      46,
      3,
      76,
      172,
      253,
      132,
      164,
      38,
      67,
      152,
      45,
      128,
      175,
      14,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      193,
      4,
      206,
      94,
      94,
      120,
      93,
      12,
      176,
      8,
      164,
      21,
      18,
      134,
      248,
      116,
      60,
      54,
      193,
      255,
      184,
      160,
      43,
      246,
      4,
      91,
      56,
      0,
      249,
      149,
      217,
      213,
      159,
      122,
      199,
      21,
      229,
      14,
      118,
      182,
      212,
      35,
      230,
      201,
      91,
      155,
      20,
      198,
      147,
      39,
      104,
      111,
      116,
      13,
      114,
      67,
      26,
      210,
      164,
      81,
      195,
      238,
      79,
      7,
    ],
    "type": "Buffer",
  },
]
`;
